### PR TITLE
Ensure that socket is closed on handshake failure

### PR DESCRIPTION
**Pull request by jyemin**
https://jira.mongodb.org/browse/JAVA-3680

I don't see a way to regression test this.  I did have the reporter run his reproducer to make sure this actually fixes the bug, and he confirmed that it did for both default and netty stream implementations.

### CODE REVIEW

File: driver-core/src/main/com/mongodb/connection/netty/NettyStream.java [hunk 1]
```diff
@@ -395,20 +395,28 @@ public T get() throws IOException {
 
         @Override
         public void operationComplete(final ChannelFuture future) {
-            if (future.isSuccess()) {
-                channel = channelFuture.channel();
-                channel.closeFuture().addListener(new ChannelFutureListener() {
-                    @Override
-                    public void operationComplete(final ChannelFuture future) {
-                        handleReadResponse(null, new IOException("The connection to the server was closed"));
+            synchronized (NettyStream.this) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/netty/NettyStream.java:398]: The root of the bug is a race condition between closing the stream, and completing the asynchronous opening of the channel.  Addressed by synchronized between the two tasks

File: driver-core/src/main/com/mongodb/connection/netty/NettyStream.java [hunk 2]
```diff
@@ -395,20 +395,28 @@ public T get() throws IOException {
...
+                if (future.isSuccess()) {
+                    if (isClosed) {
+                        channelFuture.channel().close();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/netty/NettyStream.java:401]: If the stream is closed, immediately close the channel

File: driver-core/src/main/com/mongodb/connection/netty/NettyStream.java [hunk 3]
```diff
@@ -395,20 +395,28 @@ public T get() throws IOException {
...
-                });
-                handler.completed(null);
-            } else {
-                if (socketAddressQueue.isEmpty()) {
-                    handler.failed(new MongoSocketOpenException("Exception opening socket", getAddress(), future.cause()));
+                    handler.completed(null);
                 } else {
-                    initializeChannel(handler, socketAddressQueue);
+                    if (isClosed) {
+                        handler.completed(null);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/netty/NettyStream.java:414]: If the stream is closed, don't continue trying with another socket address

File: driver-core/src/main/com/mongodb/internal/connection/AsynchronousChannelStream.java [hunk 4]
```diff
@@ -69,13 +69,17 @@ public BufferProvider getBufferProvider() {
         return bufferProvider;
     }
 
-    public ExtendedAsynchronousByteChannel getChannel() {
+    public synchronized ExtendedAsynchronousByteChannel getChannel() {
         return channel;
     }
 
-    protected void setChannel(final ExtendedAsynchronousByteChannel channel) {
+    protected synchronized void setChannel(final ExtendedAsynchronousByteChannel channel) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/AsynchronousChannelStream.java:76]: Same issue in the other async stream implementation

File: driver-core/src/main/com/mongodb/connection/netty/NettyStream.java (hunk 1 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/connection/netty/NettyStream.java:398]: Ack

**Pull request merged**